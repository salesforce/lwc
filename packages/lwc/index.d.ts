/*
 * Copyright (c) 2023, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

/**
 * Lightning Web Components core module
 */

declare module 'lwc' {
    // backwards compatible type used for the old days when TS didn't support `event.composed`
    interface ComposableEvent extends Event {
        composed: boolean;
    }

    class HTMLElementTheGoodPart {
        dispatchEvent(evt: Event): boolean;
        addEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | AddEventListenerOptions
        ): void;
        removeEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | EventListenerOptions
        ): void;
        setAttributeNS(ns: string, attrName: string, value: any): void;
        removeAttributeNS(ns: string, attrName: string): void;
        removeAttribute(attrName: string): void;
        setAttribute(attrName: string, value: any): void;
        getAttribute(attrName: string): string | null;
        getAttributeNS(ns: string, attrName: string): string | null;
        getBoundingClientRect(): ClientRect;
        querySelector<E extends Element = Element>(selectors: string): E | null;
        querySelectorAll<E extends Element = Element>(selectors: string): NodeListOf<E>;
        getElementsByTagName(tagNameOrWildCard: string): HTMLCollectionOf<Element>;
        getElementsByClassName(names: string): HTMLCollectionOf<Element>;
        attachInternals(): ElementInternals;
        readonly tagName: string;
        readonly classList: DOMTokenList;

        // Default HTML Properties
        dir: string;
        id: string;
        accessKey: string;
        title: string;
        lang: string;
        hidden: boolean;
        draggable: boolean;
        tabIndex: number;

        // Aria Properties
        ariaAutoComplete: string | null;
        ariaChecked: string | null;
        ariaCurrent: string | null;
        ariaDisabled: string | null;
        ariaExpanded: string | null;
        ariaHasPopup: string | null;
        ariaHidden: string | null;
        ariaInvalid: string | null;
        ariaLabel: string | null;
        ariaLevel: string | null;
        ariaMultiLine: string | null;
        ariaMultiSelectable: string | null;
        ariaOrientation: string | null;
        ariaPressed: string | null;
        ariaReadOnly: string | null;
        ariaRequired: string | null;
        ariaSelected: string | null;
        ariaSort: string | null;
        ariaValueMax: string | null;
        ariaValueMin: string | null;
        ariaValueNow: string | null;
        ariaValueText: string | null;
        ariaLive: string | null;
        ariaRelevant: string | null;
        ariaAtomic: string | null;
        ariaBusy: string | null;
        ariaActiveDescendant: string | null;
        ariaControls: string | null;
        ariaDescribedBy: string | null;
        ariaFlowTo: string | null;
        ariaLabelledBy: string | null;
        ariaOwns: string | null;
        ariaPosInSet: string | null;
        ariaSetSize: string | null;
        ariaColCount: string | null;
        ariaColIndex: string | null;
        ariaColIndexText: string | null;
        ariaDescription: string | null;
        ariaDetails: string | null;
        ariaErrorMessage: string | null;
        ariaKeyShortcuts: string | null;
        ariaModal: string | null;
        ariaPlaceholder: string | null;
        ariaRoleDescription: string | null;
        ariaRowCount: string | null;
        ariaRowIndex: string | null;
        ariaRowIndexText: string | null;
        ariaRowSpan: string | null;
        ariaColSpan: string | null;
        ariaBrailleLabel: string | null;
        ariaBrailleRoleDescription: string | null;
        role: string | null;
    }

    // @ts-ignore type-mismatch
    interface ShadowRootTheGoodPart extends NodeSelector {
        mode: string;
        readonly activeElement: Element | null;
        readonly host: HTMLElement;
        readonly firstChild: Node | null;
        readonly lastChild: Node | null;
        readonly innerHTML: string;
        readonly textContent: string;
        readonly childNodes: Node[];
        readonly delegatesFocus: boolean;
        addEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | AddEventListenerOptions
        ): void;
        removeEventListener(
            type: string,
            listener: EventListenerOrEventListenerObject,
            options?: boolean | EventListenerOptions
        ): void;
        hasChildNodes(): boolean;
        compareDocumentPosition(otherNode: Node): number;
        contains(otherNode: Node): boolean;
        elementFromPoint(x: number, y: number): Element | null;
        querySelector<K extends keyof HTMLElementTagNameMap>(
            selectors: K
        ): HTMLElementTagNameMap[K] | null;
        querySelector<K extends keyof SVGElementTagNameMap>(
            selectors: K
        ): SVGElementTagNameMap[K] | null;
        querySelector<E extends Element = Element>(selectors: string): E | null;
        querySelectorAll<K extends keyof HTMLElementTagNameMap>(
            selectors: K
        ): NodeListOf<HTMLElementTagNameMap[K]>;
        querySelectorAll<K extends keyof SVGElementTagNameMap>(
            selectors: K
        ): NodeListOf<SVGElementTagNameMap[K]>;
        querySelectorAll<E extends Element = Element>(selectors: string): NodeListOf<E>;
    }

    /**
     * Base class for the Lightning Web Component JavaScript class
     */
    export class LightningElement extends HTMLElementTheGoodPart {
        /**
         * This static getter builds a Web Component class from a LWC constructor so it can be registered
         * as a new element via customElements.define() at any given time. For example:
         *
         * ```
         * import XComponent from 'namespace/element';
         * customElements.define('x-component', XComponent.CustomElementConstructor);
         * const elm = document.createElement('x-component');
         * ```
         */
        static get CustomElementConstructor(): typeof HTMLElement;
        /**
         * Called when the element is inserted in a document
         */
        connectedCallback(): void;
        /**
         * Called when the element is removed from a document
         */
        disconnectedCallback(): void;
        /**
         * Called after every render of the component
         */
        renderedCallback(): void;
        /**
         * Called when a descendant component throws an error in one of its lifecycle hooks
         */
        errorCallback(error: Error, stack: string): void;

        readonly template: ShadowRootTheGoodPart;
        readonly shadowRoot: null;
        readonly refs: { [key: string]: Element | LightningElement };
    }

    /**
     * Decorator to mark public reactive properties
     */
    export const api: PropertyDecorator;

    /**
     * Decorator to mark private reactive properties
     */
    export const track: PropertyDecorator;

    type StringKeyedRecord = Record<string, any>;
    /**
     * Decorator factory to wire a property or method to a wire adapter data source
     * Use generic types to allow type checking for wire adapters
     * Default all the generic types to any to maintain backward compatibility
     *
     * For example, a wire adapter 'getRecord' can have the following type definition
     *
     * export const getRecord: WireAdapterConstructor<GetRecordConfig, RecordRepresentation>;
     *
     * in which 'GetRecordConfig' is the adapter config object type and 'RecordRepresentation'
     * is the returned value.
     *
     * @param adapter the adapter used to provision data
     * @param config configuration object for the adapter
     */
    export function wire<
        Config extends StringKeyedRecord,
        Value,
        Context extends StringKeyedRecord = StringKeyedRecord
    >(
        adapter:
            | WireAdapterConstructor<Config, Value, Context>
            | LegacyWireAdapterConstructor<Config, Value>,
        config?: WireConfigValue<Config>
    ): PropertyDecorator;

    type LegacyWireAdapterConstructor<Config, Value> = (config?: Config) => Value;
    type WireConfigValue<Config extends StringKeyedRecord = StringKeyedRecord> = {
        // wire reactive variables are strings prefixed with '$' so the config value can just be string
        [K in keyof Config]: Config[K] | string;
    };
    type ContextValue<Context extends StringKeyedRecord = StringKeyedRecord> = Context;

    interface WireAdapter<Config extends StringKeyedRecord, Context extends StringKeyedRecord> {
        update(config: WireConfigValue<Config>, context?: ContextValue<Context>): void;
        connect(): void;
        disconnect(): void;
    }

    type WireDataCallback<Value> = (value: Value) => void;
    type WireAdapterSchemaValue = 'optional' | 'required';

    interface ContextConsumer<Context extends StringKeyedRecord> {
        provide(newContext: ContextValue<Context>): void;
    }

    interface ContextProviderOptions<Context extends StringKeyedRecord> {
        consumerConnectedCallback: (consumer: ContextConsumer<Context>) => void;
        consumerDisconnectedCallback?: (consumer: ContextConsumer<Context>) => void;
    }

    interface WireAdapterConstructor<
        Config extends StringKeyedRecord,
        Value,
        Context extends StringKeyedRecord
    > {
        new (callback: WireDataCallback<Value>): WireAdapter<Config, Context>;
        configSchema?: Record<keyof Config, WireAdapterSchemaValue>;
        contextSchema?: Record<keyof Context, WireAdapterSchemaValue>;
    }

    type Contextualizer<Context extends StringKeyedRecord> = (
        elm: EventTarget,
        options: ContextProviderOptions<Context>
    ) => void;
    export function createContextProvider<
        Config extends StringKeyedRecord,
        Value,
        Context extends StringKeyedRecord = StringKeyedRecord
    >(config: WireAdapterConstructor<Config, Value, Context>): Contextualizer<Context>;
}
