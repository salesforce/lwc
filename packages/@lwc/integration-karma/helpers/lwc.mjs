import path from 'node:path';
import { rollup } from 'rollup';
import lwcRollupPlugin from '@lwc/rollup-plugin';

import {
    DISABLE_SYNTHETIC_SHADOW_SUPPORT_IN_COMPILER,
    API_VERSION,
    DISABLE_STATIC_CONTENT_OPTIMIZATION,
} from './options.mjs';

// Cache reused between each compilation to speed up the compilation time.
let cache;

export default async (ctx) => {
    const input = ctx.path.slice(1); // strip leading / from URL path to get relative file path
    const suiteDir = path.dirname(input);

    // TODO [#3370]: remove experimental template expression flag
    const experimentalComplexExpressions = suiteDir.includes('template-expressions');

    const createRollupPlugin = (options) => {
        return lwcRollupPlugin({
            // Sourcemaps don't work with Istanbul coverage
            sourcemap: !process.env.COVERAGE,
            experimentalDynamicComponent: {
                loader: 'test-utils',
                strict: true,
            },
            enableDynamicComponents: true,
            enableLwcOn: true,
            experimentalComplexExpressions,
            enableStaticContentOptimization: !DISABLE_STATIC_CONTENT_OPTIMIZATION,
            disableSyntheticShadowSupport: DISABLE_SYNTHETIC_SHADOW_SUPPORT_IN_COMPILER,
            apiVersion: API_VERSION,
            ...options,
        });
    };

    const defaultRollupPlugin = createRollupPlugin();

    const customLwcRollupPlugin = {
        ...defaultRollupPlugin,
        transform(src, id) {
            let rollupPluginToUse;

            // Override the LWC Rollup plugin to specify different options based on file name patterns.
            // This allows us to alter the API version or other compiler props on a filename-only basis.
            const apiVersion = id.match(/useApiVersion(\d+)/)?.[1];
            const nativeOnly = /\.native-only\./.test(id);
            if (apiVersion) {
                rollupPluginToUse = createRollupPlugin({
                    apiVersion: parseInt(apiVersion, 10),
                });
            } else if (nativeOnly) {
                rollupPluginToUse = createRollupPlugin({ disableSyntheticShadowSupport: true });
            } else {
                rollupPluginToUse = defaultRollupPlugin;
            }
            return rollupPluginToUse.transform.call(this, src, id);
        },
    };

    const plugins = [customLwcRollupPlugin];

    const bundle = await rollup({
        input,
        plugins,
        cache,

        // Rollup should not attempt to resolve the engine and the test utils, Karma takes care of injecting it
        // globally in the page before running the tests.
        external: ['lwc', 'wire-service', 'test-utils', '@test/loader'],

        onwarn(warning, warn) {
            // Ignore warnings from our own Rollup plugin
            if (warning.plugin !== 'rollup-plugin-lwc-compiler') {
                warn(warning);
            }
        },
    });

    cache = bundle.cache;

    const { output } = await bundle.generate({
        format: 'esm',
        // TODO: Does web-test-runner use istanbul?
        // Sourcemaps don't work with Istanbul coverage
        sourcemap: process.env.COVERAGE ? false : 'inline',

        // The engine and the test-utils is injected as UMD. This mapping defines how those modules can be
        // referenced from the window object.
        globals: {
            lwc: 'LWC',
            'wire-service': 'WireService',
            'test-utils': 'TestUtils',
        },
    });

    const { code } = output[0];
    return code;
};
