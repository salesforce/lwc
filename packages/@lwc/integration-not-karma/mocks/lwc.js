// IMPORTANT: we must use @lwc/engine-dom instead of lwc in order to avoid circular imports
import { sanitizeAttribute as _sanitizeAttribute } from '@lwc/engine-dom';
import { fn } from '@vitest/spy';

export * from '@lwc/engine-dom';
export const sanitizeAttribute = fn(_sanitizeAttribute);
