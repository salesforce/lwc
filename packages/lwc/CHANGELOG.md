# Changelog

## [7.2.3](https://github.com/salesforce/lwc/compare/v7.2.2...v7.2.3) (2024-08-30)


### Dependencies

* The following workspace dependencies were updated
  * dependencies
    * @lwc/compiler bumped from 7.2.2 to 7.2.3
    * @lwc/engine-server bumped from 7.2.2 to 7.2.3
    * @lwc/rollup-plugin bumped from 7.2.2 to 7.2.3
    * @lwc/template-compiler bumped from 7.2.2 to 7.2.3
