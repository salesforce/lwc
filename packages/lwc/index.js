/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const fs = require('fs');
const path = require('path');

const DEBUG_SUFFIX = '_debug';
const PROD_SUFFIX = '.min';

function getVersion() {
    const pkgPath = path.join(__dirname, './package.json');
    const pkgSrc = fs.readFileSync(pkgPath, 'utf8');
    const pkg = JSON.parse(pkgSrc);
    return pkg.version;
}

const MODULES = new Set(['engine-dom', 'engine-server', 'synthetic-shadow', 'wire-service']);
const FORMATS = new Set(['iife', 'umd', 'esm']);
const TARGETS = new Set(['es5', 'es2017']);
const MODES = new Set(['dev', 'prod', 'prod_debug']);

const DEFAULT_FORMAT = 'esm';
const DEFAULT_TARGET = 'es2017';
const DEFAULT_MODE = 'dev';

function validateArgs(name, format, target, mode) {
    if (!MODULES.has(name)) {
        throw new Error(
            `Invalid module name "${name}". Available module names: ${Array.from(MODULES.keys())}`
        );
    }

    if (!FORMATS.has(format)) {
        throw new Error(
            `Invalid format "${name}". Available formats: ${Array.from(FORMATS.keys())}`
        );
    }

    if (!TARGETS.has(target)) {
        throw new Error(
            `Invalid target "${target}". Available targets: ${Array.from(TARGETS.keys())}`
        );
    }

    if (!MODES.has(mode)) {
        throw new Error(`Invalid mode "${mode}". Available modes: ${Array.from(MODES.keys())}`);
    }
}

function getModule(name, format = DEFAULT_FORMAT, target = DEFAULT_TARGET, mode = DEFAULT_MODE) {
    const distPath = getModulePath(name, format, target, mode);
    return fs.readFileSync(distPath, 'utf8');
}

function getModulePath(
    name,
    format = DEFAULT_FORMAT,
    target = DEFAULT_TARGET,
    mode = DEFAULT_MODE
) {
    validateArgs(name, format, target, mode);
    let distPath;
    // The default is on the package itself (naming convention is set)
    if (format === DEFAULT_FORMAT && target === DEFAULT_TARGET && mode === DEFAULT_MODE) {
        distPath = require.resolve(`@lwc/${name}/dist/${name}.js`);

        // Otherwise is on dist of this package
    } else if (mode === 'prod') {
        distPath = path.join(__dirname, 'dist', name, format, target, `${name}${PROD_SUFFIX}.js`);
    } else if (mode === 'prod_debug') {
        distPath = path.join(__dirname, 'dist', name, format, target, `${name}${DEBUG_SUFFIX}.js`);
    } else {
        distPath = path.join(__dirname, 'dist', name, format, target, `${name}.js`);
    }

    if (!fs.existsSync(distPath)) {
        throw new Error(`Module path "${distPath}" for module ${name} not found`);
    }

    return distPath;
}

module.exports = {
    getVersion,
    getModule,
    getModulePath,
};
