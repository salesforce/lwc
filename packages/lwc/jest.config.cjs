/*
 * Copyright (c) 2024, Salesforce, Inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/* eslint-env node */

/**
 * This is a .cjs file because this package is an ESM package, but jest needs CJS to work.
 * If this file is ever changed to a .js file, add an exclusion to the package.json "files" list,
 * so that this is not shipped as part of the package.
 */

// eslint-disable-next-line @typescript-eslint/no-var-requires
const BASE_CONFIG = require('../../scripts/jest/base.config');

module.exports = {
    ...BASE_CONFIG,
    displayName: 'lwc',
    testEnvironment: 'jsdom',
};
