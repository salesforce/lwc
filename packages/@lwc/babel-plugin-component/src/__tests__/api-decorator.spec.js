/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const pluginTest = require('./utils/test-transform').pluginTest(require('../index'));

describe('Transform property', () => {
    pluginTest(
        'transforms public props',
        `
        import { api } from 'lwc';
        export default class Test {
            @api test = 1;
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Test {
                  constructor() {
                    this.test = 1;
                  }
                }

                _registerDecorators(Test, {
                  publicProps: {
                    test: {
                      config: 0
                    }
                  }
                });

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
`,
            },
        }
    );

    pluginTest(
        'transform nested classes',
        `
        import { api } from 'lwc';
        export default class Outer {
            @api outer;
            a = class {
                @api innerA;
            }
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators2 } from "lwc";
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Outer {
                  constructor() {
                    this.outer = void 0;
                    this.a = _registerDecorators2(
                      class {
                        constructor() {
                          this.innerA = void 0;
                        }
                      },
                      {
                        publicProps: {
                          innerA: {
                            config: 0
                          }
                        }
                      }
                    );
                  }
                }

                _registerDecorators(Outer, {
                  publicProps: {
                    outer: {
                      config: 0
                    }
                  }
                });

                export default _registerComponent(Outer, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'transforms public getters',
        `
        import { api } from 'lwc';
        export default class Test {
            @api get publicGetter() {
                return 1;
            }
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Test {
                  get publicGetter() {
                    return 1;
                  }
                }

                _registerDecorators(Test, {
                  publicProps: {
                    publicGetter: {
                      config: 1
                    }
                  }
                });

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'detecting @api on both getter and a setter should produce an error',
        `
        import { api } from 'lwc';
        export default class Test {
            @api
            get something() {
                return this.s;
            }
            @api
            set something (value) {
                this.s = value;
            }
        }
    `,
        {
            error: {
                message: '@api get something and @api set something',
                loc: {
                    line: 7,
                    column: 0,
                    length: 46,
                    start: 95,
                },
            },
        }
    );

    pluginTest(
        'single @api getter on one class member should not conflict with @api setter on another',
        `
        import { api } from 'lwc';
        export default class Test {
            @api
            get first() { return null; }

            get second() {
                return this.s;
            }
            @api
            set second(value) {
                this.s = value;
            }
        }
    `,
        {
            output: {
                code: `
                    import { registerDecorators as _registerDecorators } from "lwc";
                    import _tmpl from "./test.html";
                    import { registerComponent as _registerComponent } from "lwc";
                    
                    class Test {
                        get first() {
                            return null;
                        }
                        
                        get second() {
                            return this.s;
                        }
                        
                        set second(value) {
                            this.s = value;
                        }
                    }
                    
                    _registerDecorators(Test, {
                    publicProps: {
                        first: {
                            config: 1
                        },
                        second: {
                            config: 3
                        }
                    }
                    });
                    
                    export default _registerComponent(Test, {
                    tmpl: _tmpl
                    });
                `,
            },
        }
    );

    pluginTest(
        '@api setter on one class member should not conflict with @api getter on another',
        `
        import { api } from 'lwc';
        export default class Test {
            @api
            set first(value) {}
            get first() {}

            @api
            get second() {
                return this.s;
            }
            set second (value) {
                this.s = value;
            }
        }
    `,
        {
            output: {
                code: `
                    import { registerDecorators as _registerDecorators } from "lwc";
                    import _tmpl from "./test.html";
                    import { registerComponent as _registerComponent } from "lwc";
                    
                    class Test {
                    set first(value) {}
                    
                    get first() {}
                    
                    get second() {
                        return this.s;
                    }
                    
                    set second(value) {
                        this.s = value;
                    }
                    }
                    
                    _registerDecorators(Test, {
                    publicProps: {
                        first: {
                        config: 3
                        },
                        second: {
                        config: 3
                        }
                    }
                    });
                    
                    export default _registerComponent(Test, {
                    tmpl: _tmpl
                    });
                `,
            },
        }
    );

    pluginTest(
        'transform pairs of setter and getter',
        `
        import { api } from 'lwc';
        export default class Test {
            _a = true;
            _b = false;

            @api get a () { return this._a; }
            set a (value) { this._a = value; }
            @api get b () { return this._b; }
            set b (value) { this._b = value; }
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Test {
                  constructor() {
                    this._a = true;
                    this._b = false;
                  }

                  get a() {
                    return this._a;
                  }

                  set a(value) {
                    this._a = value;
                  }

                  get b() {
                    return this._b;
                  }

                  set b(value) {
                    this._b = value;
                  }
                }

                _registerDecorators(Test, {
                  publicProps: {
                    a: {
                      config: 3
                    },
                    b: {
                      config: 3
                    }
                  }
                });

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        `transform complex attributes`,
        `
        import { api } from 'lwc';
        export default class Text {
            @api publicProp;
            privateProp;

            @api get aloneGet(){}
            @api get myget(){}
            set myget(x){ return 1; }
            @api m1(){}
            m2(){}
            static ctor = "ctor";
            static get ctorGet () { return 1}
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Text {
                  constructor() {
                    this.publicProp = void 0;
                    this.privateProp = void 0;
                  }

                  get aloneGet() {}

                  get myget() {}

                  set myget(x) {
                    return 1;
                  }

                  m1() {}

                  m2() {}

                  static get ctorGet() {
                    return 1;
                  }
                }

                Text.ctor = "ctor";

                _registerDecorators(Text, {
                  publicProps: {
                    publicProp: {
                      config: 0
                    },
                    aloneGet: {
                      config: 1
                    },
                    myget: {
                      config: 3
                    }
                  },
                  publicMethods: ["m1"]
                });

                export default _registerComponent(Text, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'throws error if default value is true',
        `
        import { api } from 'lwc';
        export default class Test {
            @api publicProp = true;
        }
    `,
        {
            error: {
                message: 'Boolean public property must default to false.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 23,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'throws error if property name is "is"',
        `
        import { api } from 'lwc';
        export default class Test {
            @api is;
        }
    `,
        {
            error: {
                message: 'Invalid property name "is". "is" is a reserved attribute.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 8,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'throws error if property name prefixed with "on"',
        `
        import { api } from 'lwc';
        export default class Test {
            @api onChangeHandler;
        }
    `,
        {
            error: {
                message:
                    'Invalid property name "onChangeHandler". Properties starting with "on" are reserved for event handlers.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 21,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'does not throw error if property name is "data"',
        `
        import { api } from 'lwc';
        export default class Test {
            @api data;
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Test {
                  constructor() {
                    this.data = void 0;
                  }
                }

                _registerDecorators(Test, {
                  publicProps: {
                    data: {
                      config: 0
                    }
                  }
                });

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'throws error if property name prefixed with "data"',
        `
        import { api } from 'lwc';
        export default class Test {
            @api dataFooBar;
        }
    `,
        {
            error: {
                message:
                    'Invalid property name "dataFooBar". Properties starting with "data" are reserved attributes.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 16,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'throws error if property name is ambiguous',
        `
        import { api } from 'lwc';
        export default class Test {
            @api tabindex;
        }
    `,
        {
            error: {
                message:
                    'Ambiguous attribute name "tabindex". "tabindex" will never be called from template because its corresponding property is camel cased. Consider renaming to "tabIndex".',
                loc: {
                    line: 3,
                    column: 0,
                    length: 14,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'throws correct error if property name is maxlength',
        `
        import { api } from 'lwc';
        export default class Test {
            @api maxlength;
        }
    `,
        {
            error: {
                message:
                    'Ambiguous attribute name "maxlength". "maxlength" will never be called from template because its corresponding property is camel cased. Consider renaming to "maxLength".',
                loc: {
                    line: 3,
                    column: 0,
                    length: 15,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'does not throw if property name prefixed with "aria"',
        `
        import { api } from 'lwc';
        export default class Test {
            @api ariaDescribedBy;
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Test {
                  constructor() {
                    this.ariaDescribedBy = void 0;
                  }
                }

                _registerDecorators(Test, {
                  publicProps: {
                    ariaDescribedBy: {
                      config: 0
                    }
                  }
                });

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'throws error if property name conflicts with disallowed global html attribute name',
        `
        import { api } from 'lwc';
        export default class Test {
            @api slot;
        }
    `,
        {
            error: {
                message: 'Invalid property name "slot". "slot" is a reserved attribute.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 10,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'throws error if property name is "part"',
        `
        import { api } from 'lwc';
        export default class Test {
            @api part;
        }
    `,
        {
            error: {
                message:
                    'Invalid property name "part". "part" is a future reserved attribute for web components.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 10,
                    start: 55,
                },
            },
        }
    );

    pluginTest(
        'throws when combined with @track',
        `
        import { api, track } from 'lwc';
        export default class Test {
            @track
            @api
            apiWithTrack = 'foo';
        }
    `,
        {
            error: {
                message: '@api method or property cannot be used with @track',
                loc: {
                    line: 4,
                    column: 0,
                    length: 4,
                    start: 69,
                },
            },
        }
    );

    pluginTest(
        'Duplicate api properties',
        `
        import { api } from 'lwc';
        export default class Text {
            @api foo = 1;
            @api foo = 2;
        }
    `,
        {
            error: {
                message: 'Duplicate @api property "foo".',
                loc: {
                    line: 4,
                    column: 0,
                    length: 4,
                    start: 69,
                },
            },
        }
    );

    pluginTest(
        'Conflicting api properties with getter/setter',
        `
        import { api } from 'lwc';
        export default class Text {
            @api foo = 1;

            _internal = 1;

            @api
            get foo() { return 'foo' };
            set foo(val) { this._internal = val };
        }
    `,
        {
            error: {
                message: 'Duplicate @api property "foo".',
                loc: {
                    line: 5,
                    column: 0,
                    length: 4,
                    start: 84,
                },
            },
        }
    );

    pluginTest(
        'Conflicting api properties with method',
        `
        import { api } from 'lwc';
        export default class Text {
            @api foo = 1;
            @api foo() { return 'foo'; }
        }
    `,
        {
            error: {
                message: 'Duplicate @api property "foo".',
                loc: {
                    line: 4,
                    column: 0,
                    length: 4,
                    start: 69,
                },
            },
        }
    );
});

describe('Transform method', () => {
    pluginTest(
        'transforms public methods',
        `
        import { api } from 'lwc';
        export default class Test {
            @api foo() {}
        }
    `,
        {
            output: {
                code: `
                import { registerDecorators as _registerDecorators } from "lwc";
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";

                class Test {
                  foo() {}
                }

                _registerDecorators(Test, {
                  publicMethods: ["foo"]
                });

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
            `,
            },
        }
    );

    pluginTest(
        'Does not allow computed api getters and setters',
        `
        import { LightningElement, api } from 'lwc';
        export default class ComputedAPIProp extends LightningElement {
            @api
            set [x](value) {}
            get [x]() {}
        }
    `,
        {
            error: {
                message: '@api cannot be applied to a computed property, getter, setter or method.',
                loc: {
                    line: 3,
                    column: 0,
                    length: 22,
                    start: 109,
                },
            },
        }
    );
});
