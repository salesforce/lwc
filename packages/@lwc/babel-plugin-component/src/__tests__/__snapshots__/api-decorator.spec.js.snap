// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform method Does not allow computed api getters and setters 1`] = `
Object {
  "column": 0,
  "length": 22,
  "line": 3,
  "start": 109,
}
`;

exports[`Transform property Conflicting api properties with getter/setter 1`] = `
Object {
  "column": 0,
  "length": 4,
  "line": 5,
  "start": 84,
}
`;

exports[`Transform property Conflicting api properties with method 1`] = `
Object {
  "column": 0,
  "length": 4,
  "line": 4,
  "start": 69,
}
`;

exports[`Transform property Duplicate api properties 1`] = `
Object {
  "column": 0,
  "length": 4,
  "line": 4,
  "start": 69,
}
`;

exports[`Transform property detecting @api on both getter and a setter should produce an error 1`] = `
Object {
  "column": 0,
  "length": 46,
  "line": 7,
  "start": 95,
}
`;

exports[`Transform property throws correct error if property name is maxlength 1`] = `
Object {
  "column": 0,
  "length": 15,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if default value is true 1`] = `
Object {
  "column": 0,
  "length": 23,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if property name conflicts with disallowed global html attribute name 1`] = `
Object {
  "column": 0,
  "length": 10,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if property name is "is" 1`] = `
Object {
  "column": 0,
  "length": 8,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if property name is "part" 1`] = `
Object {
  "column": 0,
  "length": 10,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if property name is ambiguous 1`] = `
Object {
  "column": 0,
  "length": 14,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if property name prefixed with "data" 1`] = `
Object {
  "column": 0,
  "length": 16,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws error if property name prefixed with "on" 1`] = `
Object {
  "column": 0,
  "length": 21,
  "line": 3,
  "start": 55,
}
`;

exports[`Transform property throws when combined with @track 1`] = `
Object {
  "column": 0,
  "length": 4,
  "line": 4,
  "start": 69,
}
`;
